src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -675,80 +649,67 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
         case app @ Apply(fun, args) =>
           val sym = fun.symbol
 
-          if (sym.isLabel) {  // jump to a label
+          if (sym.isLabel) { // jump to a label
             genLoadLabelArguments(args, labelDef(sym), app.pos)
             bc goTo programPoint(sym)
           } else if (isPrimitive(sym)) { // primitive method call
             generatedType = genPrimitiveOp(app, expectedType)
-          } else {  // normal method call
-
-            def genNormalMethodCall() {
-
-              val invokeStyle =
-                if (sym.isStaticMember) InvokeStyle.Static
-                else if (sym.isPrivate || sym.isClassConstructor) InvokeStyle.Special
-                else InvokeStyle.Virtual
-
-              if (invokeStyle.hasInstance) {
-                genLoadQualifier(fun)
+          } else { // normal method call
+            val invokeStyle =
+              if (sym.isStaticMember) InvokeStyle.Static
+              else if (sym.isPrivate || sym.isClassConstructor) InvokeStyle.Special
+              else InvokeStyle.Virtual
+
+            if (invokeStyle.hasInstance) genLoadQualifier(fun)
+            genLoadArguments(args, paramTKs(app))
+
+            val Select(qual, _) = fun // fun is a Select, also checked in genLoadQualifier
+            if (sym == definitions.Array_clone) {
+              // Special-case Array.clone, introduced in 36ef60e. The goal is to generate this
+              // call as "[I.clone" instead of "java/lang/Object.clone" (which would also work).
+              // Not sure why this is useful, maybe for performance. The commit message says
```

Comment:
I found this part of the JVM spec hard to follow. `clone` is overriden as public in each of the `Array` classes (which also implement `Clonable`). Seems like its fine to refer the the protected method in `Object` in the invokevirtual in this case.

Toying around with this via reflection:

```
scala> java.lang.invoke.MethodHandles.publicLookup.findVirtual(classOf[Object], "clone", java.lang.invoke.MethodType.methodType(classOf[Object]))
java.lang.IllegalAccessException: member is protected: java.lang.Object.clone()Object/invokeVirtual, from java.lang.Object/public
  at java.lang.invoke.MemberName.makeAccessException(MemberName.java:852)

scala> def lookupOf(cls: Class[_]) = { import java.lang.invoke._; val constructor = classOf[MethodHandles.Lookup].getDeclaredConstructors.head; constructor.setAccessible(true); constructor.newInstance(cls).asInstanceOf[MethodHandles.Lookup] }
lookupOf: (cls: Class[_])java.lang.invoke.MethodHandles.Lookup

scala> lookupOf(classOf[Array[Int]]).findVirtual(classOf[Object], "clone", java.lang.invoke.MethodType.methodType(classOf[Object]))
res4: java.lang.invoke.MethodHandle = MethodHandle(int[])Object
```

Perhaps related:

https://bugs.openjdk.java.net/browse/JDK-4750641
https://bugs.openjdk.java.net/browse/JDK-4962591